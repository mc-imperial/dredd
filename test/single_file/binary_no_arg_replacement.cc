int main() {
  int x = 0 - 7;
  int y = 5 / 1;
  int z = -1 * 3;
}
