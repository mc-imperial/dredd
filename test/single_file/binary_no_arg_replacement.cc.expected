#include <cinttypes>
#include <cstddef>
#include <functional>
#include <string>


#ifdef _MSC_VER
#define thread_local __declspec(thread)
#elif __APPLE__
#define thread_local __thread
#endif

static thread_local bool __dredd_some_mutation_enabled = true;
static bool __dredd_enabled_mutation(int local_mutation_id) {
  static thread_local bool initialized = false;
  static thread_local uint64_t enabled_bitset[1];
  if (!initialized) {
    bool some_mutation_enabled = false;
    const char* dredd_environment_variable = std::getenv("DREDD_ENABLED_MUTATION");
    if (dredd_environment_variable != nullptr) {
      std::string contents(dredd_environment_variable);
      while (true) {
        size_t pos = contents.find(",");
        std::string token = (pos == std::string::npos ? contents : contents.substr(0, pos));
        if (!token.empty()) {
          int value = std::stoi(token);
          int local_value = value - 0;
          if (local_value >= 0 && local_value < 50) {
            enabled_bitset[local_value / 64] |= (1 << (local_value % 64));
            some_mutation_enabled = true;
          }
        }
        if (pos == std::string::npos) {
          break;
        }
        contents.erase(0, pos + 1);
      }
    }
    initialized = true;
    __dredd_some_mutation_enabled = some_mutation_enabled;
  }
  return (enabled_bitset[local_mutation_id / 64] & (1 << (local_mutation_id % 64))) != 0;
}

#define MUTATION_RETURN(arg) arg
#define REPLACE_EXPR_INT_ZERO(mutation_id_offset) if (__dredd_enabled_mutation(local_mutation_id + mutation_id_offset)) return 0
#define REPLACE_EXPR_INT_ONE(mutation_id_offset) if (__dredd_enabled_mutation(local_mutation_id + mutation_id_offset)) return 1
#define REPLACE_EXPR_INT_MINUS_ONE(mutation_id_offset) if (__dredd_enabled_mutation(local_mutation_id + mutation_id_offset)) return -1
#define REPLACE_BINARY_Sub(mutation_id_offset) if (__dredd_enabled_mutation(local_mutation_id + mutation_id_offset)) return arg1 - arg2
#define REPLACE_BINARY_Rem(mutation_id_offset) if (__dredd_enabled_mutation(local_mutation_id + mutation_id_offset)) return arg1 % arg2
#define REPLACE_BINARY_Div(mutation_id_offset) if (__dredd_enabled_mutation(local_mutation_id + mutation_id_offset)) return arg1 / arg2
#define REPLACE_BINARY_Add(mutation_id_offset) if (__dredd_enabled_mutation(local_mutation_id + mutation_id_offset)) return arg1 + arg2
#define REPLACE_BINARY_ARG2(mutation_id_offset) if (__dredd_enabled_mutation(local_mutation_id + mutation_id_offset)) return arg2
#define REPLACE_BINARY_ARG1(mutation_id_offset) if (__dredd_enabled_mutation(local_mutation_id + mutation_id_offset)) return arg1
#define MUTATION_EXPR_NOT(mutation_id_offset) if (__dredd_enabled_mutation(local_mutation_id + mutation_id_offset)) return ~(arg)
#define MUTATION_EXPR_MINUS(mutation_id_offset) if (__dredd_enabled_mutation(local_mutation_id + mutation_id_offset)) return -(arg)
#define MUTATION_PRELUDE(arg) if (!__dredd_some_mutation_enabled) return arg
static int __dredd_replace_expr_int_zero(int arg, int local_mutation_id) {
  MUTATION_PRELUDE(arg);
  REPLACE_EXPR_INT_ONE(0);
  REPLACE_EXPR_INT_MINUS_ONE(1);
  return MUTATION_RETURN(arg);
}

static int __dredd_replace_expr_int_one(int arg, int local_mutation_id) {
  MUTATION_PRELUDE(arg);
  MUTATION_EXPR_NOT(0);
  REPLACE_EXPR_INT_ZERO(1);
  REPLACE_EXPR_INT_MINUS_ONE(2);
  return MUTATION_RETURN(arg);
}

static int __dredd_replace_expr_int_minus_one(int arg, int local_mutation_id) {
  MUTATION_PRELUDE(arg);
  REPLACE_EXPR_INT_ZERO(0);
  REPLACE_EXPR_INT_ONE(1);
  return MUTATION_RETURN(arg);
}

static int __dredd_replace_expr_int_constant(int arg, int local_mutation_id) {
  MUTATION_PRELUDE(arg);
  MUTATION_EXPR_NOT(0);
  MUTATION_EXPR_MINUS(1);
  REPLACE_EXPR_INT_ZERO(2);
  REPLACE_EXPR_INT_ONE(3);
  REPLACE_EXPR_INT_MINUS_ONE(4);
  return MUTATION_RETURN(arg);
}

static int __dredd_replace_binary_operator_Sub_arg1_int_arg2_int_lhs_zero(int arg1, int arg2, int local_mutation_id) {
  MUTATION_PRELUDE(arg1 - arg2);
  REPLACE_BINARY_ARG2(0);
  return MUTATION_RETURN(arg1 - arg2);
}

static int __dredd_replace_binary_operator_Mul_arg1_int_arg2_int_lhs_minus_one(int arg1, int arg2, int local_mutation_id) {
  MUTATION_PRELUDE(arg1 * arg2);
  REPLACE_BINARY_Add(0);
  REPLACE_BINARY_Div(1);
  REPLACE_BINARY_Rem(2);
  REPLACE_BINARY_Sub(3);
  REPLACE_BINARY_ARG2(4);
  return MUTATION_RETURN(arg1 * arg2);
}

static int __dredd_replace_binary_operator_Div_arg1_int_arg2_int_rhs_one(int arg1, int arg2, int local_mutation_id) {
  MUTATION_PRELUDE(arg1 / arg2);
  REPLACE_BINARY_Add(0);
  REPLACE_BINARY_Rem(1);
  REPLACE_BINARY_Sub(2);
  REPLACE_BINARY_ARG1(3);
  return MUTATION_RETURN(arg1 / arg2);
}

int main() {
  int x = __dredd_replace_expr_int_constant(__dredd_replace_binary_operator_Sub_arg1_int_arg2_int_lhs_zero(__dredd_replace_expr_int_zero(0, 0) , __dredd_replace_expr_int_constant(7, 2), 7), 8);
  int y = __dredd_replace_expr_int_constant(__dredd_replace_binary_operator_Div_arg1_int_arg2_int_rhs_one(__dredd_replace_expr_int_constant(5, 13) , __dredd_replace_expr_int_one(1, 18), 21), 25);
  int z = __dredd_replace_expr_int_constant(__dredd_replace_binary_operator_Mul_arg1_int_arg2_int_lhs_minus_one(__dredd_replace_expr_int_minus_one(-__dredd_replace_expr_int_one(1, 30), 33) , __dredd_replace_expr_int_constant(3, 35), 40), 45);
}
